File: src/main/kotlin/taskmanager/javaway/TaskManager.java
package taskmanager.javaway;

import java.time.Instant;
import java.util.*;

public class TaskManager {
    private final Map<Integer, Task> taskStorage = new HashMap<>();

    public void addTask(Task task) {
        taskStorage.put(task.taskId(), task);
    }

    public List<Task> getAllTasks() {
        return new ArrayList<>(taskStorage.values());
    }

    public void completeTask(int taskId) {
        Task taskToComplete = findTaskById(taskId).orElseThrow();

        if (taskToComplete.isCompleted()) {
            throw new IllegalArgumentException(String.format("Task with id %d already completed!", taskId));
        }

        Task completedTask = new Task(taskToComplete.taskId(),
            taskToComplete.title(),
            true,
            taskToComplete.taskType(),
            taskToComplete.createdAt(),
            Instant.now());
        taskStorage.put(taskId, completedTask);
    }

    public void removeTask(int taskId) {
        taskStorage.remove(taskId);
    }

    public Optional<Task> findTaskById(int taskId) {
        return Optional.of(taskStorage.get(taskId));
    }

    public List<Task> filterByType(TaskType taskType) {
        return taskStorage.values().stream()
            .filter(task -> task.taskType() == taskType)
            .toList();
    }
}

File: src/main/kotlin/kotlinway/TaskManager.kt
package ru.sinvic.kotlinway

class TaskManager {
    private val taskStorage = HashMap<Int, Task>();

    operator fun plusAssign(task: Task) {
        taskStorage[task.id] = task
    }

    operator fun minusAssign(id: Int) {
        taskStorage.remove(id)
    }

    fun completeTask(id: Int) {
        findTaskById(id)?.let { task ->
            require(!task.isCompleted) { "Task with ID=$id is already completed" }
            taskStorage[id] = task.copy(isCompleted = true)
        } ?: throw NoSuchElementException("Task with ID=$id is not found")
    }

    fun findTaskById(id: Int): Task? = taskStorage[id]

    fun getAllTasks(): List<Task> = taskStorage.values.toList()

    fun getTaskByType(taskType: TaskType): List<Task> = taskStorage.values.byType(taskType)

    private fun Collection<Task>.byType(type: TaskType): List<Task> =
        filter { it.type == type }
}
